### ISSUE TITLE
Cactbot not working over WebSocket externally

### ISSUE THREAD

**Issue by valarnin** [quisquous/cactbot#1079]
## Description

Discovered this issue alongside #1078, Cactbot doesn't seem to get a response for `callOverlayHandler` from `raidboss.js` when connected to externally, seems to work fine internally.

## Additional information

No error message for this that I've seen, just never seems to get a response.

This URL works:
```
http://192.168.1.111:38080/Cactbot/ui/raidboss/raidboss.html?OVERLAY_WS=ws://192.168.1.111:10501/ws
```

This URL doesn't work:
```
http://<external IP>:38080/Cactbot/ui/raidboss/raidboss.html?OVERLAY_WS=ws://<external IP>:10501/ws
```

However, this URL works for DPS overlay, so the issue isn't with port forwarding or anything.
```
http://<external IP>:38080/mopimopi/?HOST_PORT=ws://<external IP>:10501/#
```

**Comment by ngld** [comment]
If I had to guess the most likely cause would be the data files handling. Do you have a log of the WebSocket traffic?

**Comment by valarnin** [comment]
Not sure how I'd go about getting a log of the websocket traffic itself. That doesn't show in Chrome's dev tools (just the initial request to `/ws` which gets a 101 response code).

**Comment by ngld** [comment]
If you select that request, you should be able to see the traffic on one of the tabs to the right.

**Comment by valarnin** [comment]
Contents of the ws traffic tab after 5 minutes (to allow for any timeouts or w/e). Formatting isn't great since the only option was to select all and copy out, but it's:
```
request/response  size
timestamp
```

```
{"call":"subscribe","events":[]}	32	
23:22:05.758
{"call":"subscribe","events":["onUserFileChanged"]}	51	
23:22:06.327
{"call":"subscribe","events":["onForceReload"]}	47	
23:22:06.327
{"call":"cactbotLoadData","overlay":"options","rseq":0}	55	
23:22:06.327
{"call":"cactbotLoadUser","source":"http://<external IP>:38080/Cactbot/ui/raidboss/raidboss.html?OVERLAY_WS=ws://<external IP>:10501/ws","rseq":1}	148	
23:22:06.327
{"data":{"jobs":{"HideWellFedAboveSeconds":"600","ShowMPTickerOutOfCombat":true}},"rseq":0}	91	
23:22:06.420
{"detail":{"userLocation":"C:\\users\\steamuser\\Application Data\\Advanced Combat Tracker\\Plugins\\Cactbot\\user","localUserFiles":{"raidboss.js":"Options.PerTriggerOptions = {\n 'TEA Limit Cut Numbers': {\n SpeechAlert: true,\n SoundAlert: false,\n TextAlert: false,\n TTSText: function(data) {\n return data.limitCutNorthSouth == 'N' ? 'North' : 'South';\n },\n PreRun: function(data, matches) {\n data.limitCutNumber = {\n '004F': 1,\n '0050': 2,\n '0051': 3,\n '0052': 4,\n '0053': 5,\n '0054': 6,\n '0055': 7,\n '0056': 8,\n }[matches.id];\n if (data.phase == 'wormhole') {\n data.limitCutDelay = {\n '004F': 9.2,\n '0050': 10.7,\n '0051': 13.4,\n '0052': 15.0,\n '0053': 17.7,\n '0054': 19.2,\n '0055': 22.0,\n '0056': 23.4,\n }[matches.id];\n data.limitCutNorthSouth = (((data.limitCutNumber - 1) % 4) < 2 ? 'N' : 'S');\n } else {\n data.limitCutDelay = {\n '004F': 9.5,\n '0050': 11,\n '0051': 14.1,\n '0052': 15.5,\n '0053': 18.6,\n '0054': 20,\n '0055': 23.2,\n '0056': 24.6,\n }[matches.id];\n }\n },\n 'AlertText': function(data) {\n if(data.phase !== 'wormhole') {\n var modPos = (data.limitCutNumber - 1) % 4;\n var enText = '#' + data.limitCutNumber + ': ' + (data.limitCutNorthSouth == 'N' ? 'North' : 'South');\n \n return {\n en: enText,\n de: '#' + data.limitCutNumber + ': ' + (data.limitCutNorthSouth == 'N' ? 'Norden' : 'Süd'),\n ja: data.limitCutNumber + '番: ' + (data.limitCutNorthSouth == 'N' ? '北' : '南'),\n };\n } else {\n //Original logic below for Wormhole\n return {\n en: '#' + data.limitCutNumber,\n de: '#' + data.limitCutNumber,\n ja: data.limitCutNumber + '番',\n };\n }\n },\n },\n};\n\n//# sourceURL=C:\\users\\steamuser\\Application Data\\Advanced Combat Tracker\\Plugins\\Cactbot\\user\\raidboss.js"},"language":"en"},"rseq":1}	2668	
23:22:06.690
{"call":"subscribe","events":["onLogEvent"]}	44	
23:22:06.691
{"call":"cactbotReadDataFiles","source":"http://<external IP>:38080/Cactbot/ui/raidboss/raidboss.html?OVERLAY_WS=ws://<external IP>:10501/ws","rseq":2}	153	
23:22:06.691
{"call":"subscribe","events":["PartyChanged"]}	46	
23:22:06.693
{"call":"subscribe","events":["onPlayerChangedEvent"]}	54	
23:22:06.693
{"call":"subscribe","events":["onZoneChangedEvent"]}	52	
23:22:06.693
{"call":"subscribe","events":["onInCombatChangedEvent"]}	56	
23:22:06.693
{"call":"cactbotRequestState","rseq":3}	39	
23:22:06.693
{"type":"PartyChanged","party":[]}	34	
23:22:06.778
{"$isNull":true,"rseq":3}	25	
23:22:07.046
{"detail":{"files":null},"rseq":2}	34	
23:23:46.931
```

**Comment by ngld** [comment]
The reply to `cactbotReadDataFiles` is `{"files":null}`. Looks like my assumption was right, it fails to load the data files.
Can you check the overlay log for errors? I think Cactbot is trying to load http://<external IP>:38080/Cactbot/ui/raidboss/data/manifest.txt which fails (probably timeout) and then causes the `null` response since it can't load the manifest.

**Comment by valarnin** [comment]
Looks like it *is* timing out in the plugin itself, but (probably due to linux) the error messages aren't visible until I resize the error log in ACT. Either way:
```
[2/23/2020 11:23:50 PM] Error: Unable to read manifest file: The operation has timed out
```

I checked the source and it looks like it's trying to make Cactbot ACT-side read the manifest from the source URL, which seems like a potential security issue but given how rare this sort of configuration likely is, I'm not too fussed about that.

The source of the problem appears to be that my router does not do hair pin NAT by default, so when ACT-side Cactbot requests the manifest file from the passed in IP which is external, it fails to connect. I've added the required rules for hair pin NAT and will re-test later today to ensure that it actually resolves the problem.
